ec_api/apps/api_proxy/auth.py

```diff
@@ -25,12 +28,40 @@ def authenticate_credentials(self, key):
         Attempt to authenticate the api key. On success returns the user and
         API key object to be added to the request.
         """
-        model = self.get_model()
-        try:
-            token = model.objects.get(key=key)
-        except model.DoesNotExist:
+
+        # First try to get the key from the cache
+        cached_user = cache.get(key)
+        if cached_user is None:
+            # We've not cached this key before. We need to see if it's valid
+            model = self.get_model()
+            try:
+                token = model.objects.get(key=key)
+                # The User is valid, so set the cache with the value being a
+                # dict with the User PK
+                cached_user = {"pk": token.user.pk}
+                cache.set(key, cached_user, 60 * 5)
+            except model.DoesNotExist:
+                # This isn't a valid key. To prevent this being a way to bruit
+                # force the database, we'll cache the key with a string that
+                # tells us it's invalid
+
+                # Cache invalid keys for 30 minutes
+                cache.set(key, self.invalid_string, 60 * 30)
+                cached_user = self.invalid_string
+
+        if cached_user == self.invalid_string:
+            # If we've seen an invalid key, fail authrntication
             raise exceptions.AuthenticationFailed("Invalid token.")
-        return (token.user, token)
+
+        # Construct a CustomUser object to return. This isn't a saved object
+        # but if the actual object is needed downstream we can either call
+        # `refresh_from_db()` on it, or add more attributes to the cached dict
+        # above
+        user_obj = CustomUser(**cached_user)
+        user_obj.auth_check_keys = False
+
+        # return the user object, and the API key
+        return user_obj, key
```

Comment:
Subsequently noticed that one of the tests is failing - I think that error is due to the cache being unable to cache a mock object, as the APIKey object is actually a mock, so `key.user.pk` will return a mock.

But also made me notice that the return value here has actually changed - previously it was returning the `APIKey` instance, now it returns the [actual string key](https://github.com/DemocracyClub/ec-api-proxy/blob/a4032f664741c2fe30ad7b61665d103523226584/ec_api/apps/users/models.py#L67) `APIKey.key`.

Honestly can't remember why the object was getting returned before, or if this is going to have a knock on elsewhere in the code somehow, but if other tests are failing I presume its fine. So I think to update the failing test you could change it to:

```
def test_authenticate_credentials_valid_key(self, mocker, api_key):
    """
    Test a CustomUser object and the key are returned when key is authenticated
    """
    key = mocker.MagicMock(spec=APIKey, key=api_key)
    key.user.pk = 1
    mocker.patch.object(APIKey.objects, "get", return_value=key)

    auth_obj = TokenWithGetParamAuthentication()
    result = auth_obj.authenticate_credentials(key=api_key)

    assert type(result[0]) == CustomUser
    assert result[1] == key.key
```